racket/collects/racket/treelist.rkt

```diff
@@ -0,0 +1,1327 @@
+#lang racket/base
+(require (for-syntax racket/base
+                     syntax/for-body)
+         (only-in '#%kernel
+                  vector-copy
+                  vector*-copy vector*-append vector*-set/copy)
+         '#%flfxnum
+         racket/hash-code
+         "private/sort.rkt")
+
+;; RRB tree implemention
+;;
+;; Based on
+;;
+;;  Phil Bagwell and Tiark Rompf, "RRB Trees: Efficient Immutable Vectors"
+;;
+;;  Nicolas Stucki, Tiark Rompf, Vlad Ureche, Phil Bagwell, "RRB-Trees: Efficient Immutable Vectors"
+;;
+;;  Jean Niklas L'orange, "Improving RRB-Tree Performance through Transience"
+;;
+;; and ported from an initial implementation in Rhombus by Alec Mills.
+
+(provide (rename-out [build-treelist treelist])
+         treelist?
+         treelist-ref
+         treelist-first
+         treelist-last
+         treelist-set
+         treelist-add
+         treelist-cons
+         treelist-append
+         treelist-insert
+         treelist-delete
+         treelist-take
+         treelist-drop
+         treelist-take-right
+         treelist-drop-right
+         treelist-sublist
+         treelist-rest
+         treelist-length
+         treelist-empty?
+         empty-treelist
+         treelist->vector
+         vector->treelist
+         treelist->list
+         list->treelist
+         treelist-map
+         treelist-for-each
+         treelist-member?
+         treelist-find
+         treelist-sort
+         in-treelist
+         for/treelist
+         for*/treelist
+         chaperone-treelist)
+
+(module+ unsafe
+  (provide
+   ;; For constructing a treelist without intermediate wrapper
+   ;; structs, instead directly wirking with the root node,
+   ;; size, and depth as separate values:
+   unsafe-root-add
+   (rename-out [empty-node unsafe-empty-root]
+               [treelist unsafe-treelist])
+   ;; For implementing mutable-treelist:
+   treelist-copy-for-mutable
+   treelist-set!
+   ;; To perform the same checks, but for mutable:
+   check-treelist-index
+   check-treelist-end-index
+   check-treelist-bound-index
+   check-sort-arguments
+   check-chaperone-properties
+   ;; For printing, comparing, etc., mutable treelists:
+   treelist-print
+   treelist-equal?
+   treelist-hash-code
+   impersonate-treelist
+   unimpersonate-treelist
+   (for-syntax make-for/treelist)))
+
+(#%declare #:unsafe)
```

Comment:
This would be the second module in `racket/collects` to use `(#%declare #:unsafe)`, following `net/base64` since https://github.com/racket/racket/commit/ef43f26f6240985385b7b62e539950b542bdda51. (Of course, many other modules use individual unsafe functions.) I don't have a particular conclusion about this, but it seemed worth noting.

More concretely, `net/base64` puts `(#%declare #:unsafe)` well within the first screenful of the file, which seems worth emulating.

Reply:
I moved `(#%declare #:unsafe)` to the top.